### ISSUE TITLE
Alert rises for a chart which no longer updates

### ISSUE THREAD

**Issue by yasharne** [netdata/netdata#8751]
##### Bug report summary
I have added a revoked certificate to test if I get the revocation alert, but after removing the certificate from `x509conf` file, I still see the chart on the dashboard and I get the alert daily (I have set the revocation alert to repeat every day)
As you can see, the latest data collection was April 12th 15:10 :
![image](https://user-images.githubusercontent.com/7656049/79580230-f54fd900-80dd-11ea-93d3-8432cb115d7d.png)
But I'm still getting the alert on April 17th :
![image](https://user-images.githubusercontent.com/7656049/79580360-1f090000-80de-11ea-8480-35b33e4ff559.png)

##### OS / Environment

```

```

##### Netdata version

 
The slave which has the X509 plugin runs Netdata V1.21.1
The Master Server which receives Stream from the slave runs Netdata 1.21.0
##### Component Name

health
##### Steps To Reproduce

1. Add a certificate to the `x509` plugin and wait for the alert
2. remove the job from the `x509`conf file and restart the Netdata

##### Expected behavior

Do not show the old chart which is removed

**Comment by thiagoftsm** [comment]
@ilyam8, please, can you see whether this is a go.d.plugin bug ?

**Comment by ilyam8** [comment]
> remove the job from the x509conf file and restart the Netdata

No, it is not. I guess it is well know problem of master slave setup.

1. Master: rec chart A, B. Apply alarm for A, B.
2. Slave: remove job, restart slave:
3. Master: rec chart A. Lack of chart B triggers alarm to raise.

**Comment by yasharne** [comment]
@ilyam8 Is there any workaround?

**Comment by ilyam8** [comment]
I think it was fixed or there is some workaround, but i am not aware of it @vlvkobal @thiagoftsm

**Comment by thiagoftsm** [comment]
Hi guys,

We are developing a working around the health now, @mfundul detected important problems that needed to be fixed there. We hope that this will fix some problems and probably this one, but case it is not solved, I will plan this for next sprints to me.

**Comment by ilyam8** [comment]
i think @mfundul fix doesnt solve the problem

@thiagoftsm i suggest to reproduce the bug, did you try it?

**Comment by thiagoftsm** [comment]
@mfundul works is related with masters that needs a bigger time to send a message, because the database was locked. 
Tomorrow I will do a test with the `X509` plugin to replicate this.

**Comment by ilyam8** [comment]
It is no `x509` specific.

@yasharne correct me if i am wrong, to reproduce it we need:

 1. setup master/slave
 2. enable module on the slave with 2 jobs (easiest way is to use `example` python.d module)
 3. create availability alarm ([example](https://github.com/netdata/netdata/blob/430c91bfc8656d57d697234468676a88ed2ca36b/health/health.d/apache.conf#L4-L13)), ensure it is applied to both jobs (check the dashboard)
 4. remove one job on the slave, restart slave

Result: raised `number of seconds since the last successful data collection` alarm for the 2nd job (which was deleted in 4 step)

**Comment by yasharne** [comment]
Exactly @ilyam8 , I had the same problem with `tcp_portcheck` plugin streaming to the master as well

**Comment by thiagoftsm** [comment]
Perfect, when I finish the tests, I will bring the results here.

**Comment by mfundul** [comment]
@thiagoftsm I don't think this bug is related to any delay in message delivery.

**Comment by mfundul** [comment]
This has to do with the overall architecture of netdata, and we should keep this case in mind as we change the architecture to allow access to more historical data.

Maybe a possible workaround is to restart the master server, but I don't think there is a fix for it.

**Comment by thiagoftsm** [comment]
I will do a deeper investigation in this case tomorrow, I will set only one master and one slave to verify whether this is also happening in the simplest case possible, right now, I am thinking it is.

**Comment by ilyam8** [comment]
This is actually a big blocker for k8s and any dynamic environment.

**Comment by yasharne** [comment]
@ilyam8 @thiagoftsm  Can I get rid of this error by deleting the database file in the master? is it enough to delete  `health-log.db` file from `/opt/netdata/var/lib/netdata/` directory? is it safe?
It's really getting annoying to get a 'test' alert every day

**Comment by thiagoftsm** [comment]
Hi @yasharne ,

The file `health-log.db` stores the logs for the last raised alarms, case you stop your master and remove it, you won't reload the same alarms again.

Unfortunately I could not recreate the same error, because I am having another in my master that I am debugging now, I will continue working with this tomorrow.

**Comment by thiagoftsm** [comment]
Hi guys,

I debugged the problem using `example.chart.py` ( Thank you @ilyam8 for the suggestion! ) and I have now a more clear vision about what is happening:

1 - The slave connected with using 5 dimensions on the chart `example.chart.py`
2 - I stopped the slave and I removed 4 dimensions.
3 - I start the slave again.

When the slave restart the master did not recreate the alarms based in the new chart dimensions, it kept the first chart definition and it continued running them.  To detect this problem I used the following api calls:

```
https://localhost:19999/host/SLAVE/api/v1/alarms?all
https://localhost:19999/host/SLAVE/api/v1/charts?all
```

**Comment by ilyam8** [comment]
> When the slave restart the master did not recreate the alarms based in the new chart dimensions

Yeah.

Netdata master should handle 2 cases:

1. netdata slave deletes charts during runtime
2. delete slave deletes charts after restart

___

I think first one is handled already. It is when a plugin sends chart definition with `obsolete` or `obsolete hidden` options. 

And that is exactly the case in `k8s`. So it is not a `k8s` blocker 🎉 

Example:
```sh
CHART 'example.random' '' 'A Random Number' 'random' 'random' '' '' '70000' '1' 'obsolete' 'go.d' 'example'
```

___

2nd case is more tricky. Netdata master should handle it itself.

If a slave was restarted master should figure out `set(old slave charts) - set(current slave chart)` and mark all the charts in there as `obsolete` (and probably `hidden`).

I am not sure it is possible since it is a stream of metrics and there is no _blocks_ (BEGIN ALL CHARTS END). 

The slave should notify the master that it was restarted on connect somehow.

**Comment by yasharne** [comment]
Is there any update on this matter? I also ran to this problem with these steps:
1. the old `python.d.conf` had this setting:
```
#
# Ansible managed
#
enabled: yes
default_run: no
portcheck: yes
redis: yes
```
I changed it with this configuration:
```
#
# Ansible managed
#
enabled: yes
default_run: no
haproxy: yes
```
and then netdata started to send alerts to our monitoring channel like this (per each portcheck and redis target):
![image](https://user-images.githubusercontent.com/7656049/80863421-615f3f00-8c91-11ea-9b59-c9e004173740.png)

**Comment by thiagoftsm** [comment]
Hi @yasharne ,

We had an internal talk and we are working for to bring a definitive solution not only for this, but to improve the Health system. 
There is a spike moving forward in this area and we hope to bring the solution ASAP.

Best regards!

**Comment by yasharne** [comment]
Hi @thiagoftsm 
I still have this problem on v1.23.0

**Comment by ilyam8** [comment]
@yasharne as a temorary workaround you can do this

[latest version](https://github.com/netdata/go.d.plugin/releases/tag/v0.19.2) support reload (no need to restart netdata if you want to reload go.d.plugin)

> sudo kill -HUP $(pidof go.d.plugin)

**Comment by thiagoftsm** [comment]
Hi @yasharne ,

I apologize the delay to answer, but I am busy with a new feature these days.
We did not merge all changes we are working during the last release. I assigned our developer that is working with the stream for him take a look on this.
@amoss , please, see the comment https://github.com/netdata/netdata/issues/8751#issuecomment-616820633 above with some details about the error that is happening with stream and health.

Best regards!

**Comment by yasharne** [comment]
@ilyam8 reloading go plugin didn't fix the problem

**Comment by ilyam8** [comment]
Well, i was thinking reload fixes it because `go.d.plugin` sends all active charts with `obsolete` flag on reload

Example with 3 example jobs running:

```cmd
[ INFO  ] main[main] plugin.go:98 received hangup signal (1), stopping running instance
[ INFO  ] discovery[file manager] discovery.go:72 instance is stopped
[ INFO  ] run[manager] run.go:31 instance is stopped
[ INFO  ] build[manager] build.go:107 instance is stopped
[ INFO  ] discovery[manager] manager.go:96 instance is stopped
CHART 'netdata.execution_time_of_example_example_local_1' '' 'Execution Time for example_example_local_1' 'ms' 'go.d' 'netdata.go_plugin_execution_time' '' '145000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'time' '' '' '1' '1' ''

CHART 'example_example_local_1.random' '' 'A Random Number' 'random' 'random' '' '' '70000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'random0' 'random 0' '' '1' '1' ''
DIMENSION 'random1' 'random 1' '' '1' '1' ''

[ INFO  ] example[example_local_1] job.go:173 stopped
CHART 'netdata.execution_time_of_example_example_local_2' '' 'Execution Time for example_example_local_2' 'ms' 'go.d' 'netdata.go_plugin_execution_time' '' '145000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'time' '' '' '1' '1' ''

CHART 'example_example_local_2.random' '' 'A Random Number' 'random' 'random' '' '' '70000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'random0' 'random 0' '' '1' '1' ''
DIMENSION 'random1' 'random 1' '' '1' '1' ''

[ INFO  ] example[example_local_2] job.go:173 stopped
CHART 'netdata.execution_time_of_example_example_local_3' '' 'Execution Time for example_example_local_3' 'ms' 'go.d' 'netdata.go_plugin_execution_time' '' '145000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'time' '' '' '1' '1' ''

CHART 'example_example_local_3.random' '' 'A Random Number' 'random' 'random' '' '' '70000' '1' 'obsolete' 'go.d' 'example'
DIMENSION 'random0' 'random 0' '' '1' '1' ''
DIMENSION 'random1' 'random 1' '' '1' '1' ''

[ INFO  ] example[example_local_3] job.go:173 stopped
[ INFO  ] main[main] plugin.go:107 instance is stopped
```

i think netdata should disable alarms for inactive charts.

**Comment by yasharne** [comment]
@ilyam8 I'm using netdata `1.23.2` with go version of `0.19.2` on the child and netdata `1.23.0` on the parent

**Comment by ilyam8** [comment]
- i created 3 exmample jobs

```yaml
jobs:
  - name: example_local_1
  - name: example_local_2
  - name: example_local_3
```

- created 3 alarms

```yaml
template: example1_last_collected_secs
      on: example_example_local_1.random
    calc: $now - $last_collected_t
   units: seconds ago
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
    crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
   delay: down 5m multiplier 1.5 max 1h
    info: number of seconds since the last successful data collection
      to: webmaster

template: example2_last_collected_secs
      on: example_example_local_2.random
    calc: $now - $last_collected_t
   units: seconds ago
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
    crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
   delay: down 5m multiplier 1.5 max 1h
    info: number of seconds since the last successful data collection
      to: webmaster

template: example3_last_collected_secs
      on: example_example_local_3.random
    calc: $now - $last_collected_t
   units: seconds ago
   every: 10s
    warn: $this > (($status >= $WARNING)  ? ($update_every) : ( 5 * $update_every))
    crit: $this > (($status == $CRITICAL) ? ($update_every) : (60 * $update_every))
   delay: down 5m multiplier 1.5 max 1h
    info: number of seconds since the last successful data collection
      to: webmaster
```

- restarted netdata

I see 3 alarms

- removed one job

```yaml
jobs:
  - name: example_local_1
  - name: example_local_2
  # - name: example_local_3
```

- reloaded go plugin

> kill -HUP $(pidof go.d.plugin)

- checked alarms, there is only 2, the one for the disabled job was turned off

**Comment by ilyam8** [comment]
I used latest master, i didnt test child/master setup tho 🤔

**Comment by ilyam8** [comment]
Well, i tested child/parent setup. I did same steps as in https://github.com/netdata/netdata/issues/8751#issuecomment-665601439

On my parent node (`http://192.168.88.250:19998/host/ilyam-pc/`) i see no `example_local_3` charts and `example3_last_collected_secs` alarm.

@yasharne 

> netdata 1.23.0 on the parent

this could be the problem

**Comment by ilyam8** [comment]
@yasharne 

> reloading go plugin didn't fix the problem

Did you check child netdata dashboard? Is removed job in there or not?

**Comment by yasharne** [comment]
in the child, `memory mode` and `web` were set to none, I enabled the web and the removed charts weren't there but I guess this was expected as well, I think the problem is in the parent. because deleting the DB engine files on the parent and health data will solve the problem

**Comment by mfundul** [comment]
> 
> 
> in the child, `memory mode` and `web` were set to none, I enabled the web and the removed charts weren't there but I guess this was expected as well, I think the problem is in the parent. because deleting the DB engine files on the parent and health data will solve the problem

`v1.23.2` should have some fixes related to health and memory mode `dbengine`.

**Comment by thiagoftsm** [comment]
Hi everyone,

I worked today on this issue and I have observed that there was progress with Netdata 1.23 when the function  rrdhost_update was created

https://github.com/netdata/netdata/blob/5625d4727f38fc1d3060ab890f18d80ea8cf7d7e/database/rrdhost.c#L396

I repeated the tests @ilyam8 suggested [here](https://github.com/netdata/netdata/issues/8751#issuecomment-615899174) and I observed that after Netdata restart, I continue having alarms set for dimensions that I disabled on `child`.  This is happening, because we are calling the function in the beginning of the streaming process instead of waiting for charts update. @amoss is already working at #9260 to bring a better synchronization between parent and child and I will have a talk with him to define the best time to reload the health.

**Comment by amoss** [comment]
That is interesting. So the call at the beginning of the stream will be the metadata reload that puts the chart into archive mode?

**Comment by thiagoftsm** [comment]
> That is interesting. So the call at the beginning of the stream will be the metadata reload that puts the chart into archive mode?

Yes, we are doing this and I do not think we need to remove the health reload from this position, because it has the potential to apply the most recent changes on charts, but we cannot discard to reload the alarms when dimensions are set as `hidden`. I am considering the following reloads for alarms:

1 - When `child` connects with parent.
2 - After the new chart definitions are received.
3 - Finally, I did not review completely the PR #9260, I only saw what you are bringing there, because it has relationship with this issue, but considering what I saw there, it is possible that when a `REPEND` will be received, we will also need to reload. I prefer do not do this affirmation yet, because I need to do a proper review first.

If we do reload in these three conditions, this issue will be completely fixed. The scenario `1`, is already done, the scenario `2` we can bring right now, but for `3` we will neeed #9260 merged.

**Comment by amoss** [comment]
> 3 - Finally, I did not review completely the PR #9260, I only saw what you are bringing there, because it has relationship with this issue, but considering what I saw there, it is possible that when a `REPEND` will be received, we will also need to reload. I prefer do not do this affirmation yet, because I need to do a proper review first.

Yes I think that you are correct. When the `REPEND` is received the state in the `RRSET` and `RRDDIM` jumps over a window of time (that matches the replication block). So each of the calculated values can change in this window.

Not all replication blocks are current - some of them are old. The processing of `REPMETA` indicates that the receiver has caught up with the current data on the sender, this might be a good time to reload the alarm states.

**Comment by amoss** [comment]
We have several related problems in this area and after discussing it with @thiagoftsm we would suggest the following solution:

* Add a new state to alarms in Netdata that is `unknown`.
* If there is no data received during the collection window for an alarm it would transit into the `unknown` state.
* New data being received from a collector would then flip from `unknown` into the correct state (clear, warning or critical).
* There would be a new notification message for the state that tells users no data has been received (and for how long).

We think this would solve several of the current issues were there is no approach that is always correct as if we are not receiving data from a collector then we do not know what the real state of the alarm is. Communicating this could help the user in two ways:

1. If the alarm stops because the service is gone then they no longer have an inaccurate active alarm.
2. The service stopping sending data could be useful information by itself, i.e. I need to log in and check something

@stelfrag This may relate to the current issue with cloud alarms and it could help if we bring the agent up with `unknown` as the default state until we receive data from a collector.

**Comment by stelfrag** [comment]
> @stelfrag This may relate to the current issue with cloud alarms and it could help if we bring the agent up with `unknown` as the default state until we receive data from a collector.

We may want to reconsider the "send the active alarms from the alarm log" altogether after all. Alarms start "uninitialized" and change state as needed :thinking:

**Comment by MrZammler** [comment]
Hi! There have been some recent changes to how a parent will handle alerts for a child who's _charts_ become obsolete, especially with the `last_collected` related alerts. Also not sure how relevant the discussion on cloud alerts still is...

@ilyam8 / @stelfrag  should we keep this discussion still ?

Also @yasharne this seems related to other tickets regarding children/parent setups and alerts. I suggest we follow the problems you are having on a single issue?

**Comment by ilyam8** [comment]
I believe the case described in "Steps To Reproduce" is fixed in the latest master (#12707). I think we can close the issue @MrZammler.

**Comment by MrZammler** [comment]
Thanks!
